db-tests/src/test/java/gov/nasa/jpl/aerie/database/ExternalEventTests.java

```diff
@@ -0,0 +1,2169 @@
+     *    testing any properties of said events.
+     */
+    @Nested
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    class DerivedSourcesTests {
+
+      // Commonly Repeated:
+      final static String DURATION = "00:00:00.000001";
+
+      @BeforeEach
+      void beforeEach() throws SQLException {
+        // insert generic external event type, source type, and derivation group
+        insertStandardTypes();
+      }
+
+      /**
+       * The first test is a basic, non-overlapping case. We must ensure that gaps are preserved:
+       *
+       *    A:             ++++++++
+       *    B:  +++++++
+       *        BBBBBBB    AAAAAAAA
+       */
+      @Test
+      void testSparseCoverage() throws SQLException {
+        // create our sources and their per-window events
+        ExternalSource a = new ExternalSource(
+            "A",
+            SOURCE_TYPE,
+            DERIVATION_GROUP,
+            "2024-01-01T00:00:00Z",
+            "2024-01-01T03:00:00Z",
+            "2024-01-01T04:00:00Z",
+            CREATED_AT
+        );
+        ExternalEvent aE = new ExternalEvent(a.key() + "_event", SOURCE_TYPE, a.key(),
+                                             DERIVATION_GROUP, a.start_time(), DURATION);
+        ExternalSource b = new ExternalSource(
+            "B",
+            SOURCE_TYPE,
+            DERIVATION_GROUP,
+            "2024-01-02T00:00:00Z",
+            "2024-01-01T01:00:00Z",
+            "2024-01-01T02:00:00Z",
+            CREATED_AT
+        );
+        ExternalEvent bE = new ExternalEvent(b.key() + "_event", SOURCE_TYPE, b.key(),
+                                             DERIVATION_GROUP, b.start_time(), DURATION);
+
+        // verify the ranges are as expected
+        // insert sources
+        insertExternalSource(a);
+        insertExternalSource(b);
+
+        // insert events
+        insertExternalEvent(aE);
+        insertExternalEvent(bE);
+
+        var results = getDerivedEvents("ORDER BY source_key");
+
+        // both ranges should only have a single element and be fully present
+        final List<DerivedEvent> expectedResults = List.of(
+            new DerivedEvent(
+                "A_event",
+                SOURCE_TYPE,
+                "A",
+                DERIVATION_GROUP,
+                "2024-01-01 03:00:00+00",
+                "00:00:00.000001",
+                "{[\"2024-01-01 03:00:00+00\",\"2024-01-01 04:00:00+00\")}",
+                "2024-01-01 00:00:00+00"
+            ),
+            new DerivedEvent(
+                "B_event",
+                SOURCE_TYPE,
+                "B",
+                DERIVATION_GROUP,
+                "2024-01-01 01:00:00+00",
+                "00:00:00.000001",
+                "{[\"2024-01-01 01:00:00+00\",\"2024-01-01 02:00:00+00\")}",
+                "2024-01-02 00:00:00+00"
+            )
+        );
+        assertTrue(results.containsAll(expectedResults));
```

Comment:
This check requires an addtional assertion before it:
```java
assertEquals(expectedResults.size, results.size)
```

By adding this check, we now know that `results` _only_ contains the expected results, and not `expectedResults` plus some other entries.

This feedback applies throughout the file.